tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers.ts

```diff
@@ -504,8 +510,90 @@ const reducer = createReducer(
             lastLoadedTimeInMs: Date.now(),
           },
           sourceFileList: sourceFileList.sourceFiles,
+          fileContents: state.sourceCode.fileContents.slice(),
+        },
+      };
+      const newNumFiles = sourceFileList.sourceFiles.length;
+      for (let i = 0; i < newNumFiles; ++i) {
+        newState.sourceCode.fileContents[i] = state.sourceCode.fileContents[
+          i
+        ] || {
+          loadState: DataLoadState.NOT_LOADED,
+          lines: null,
+        };
+      }
+      return newState;
+    }
+  ),
+  on(
+    actions.sourceLineFocused,
+    (state: DebuggerState, focus): DebuggerState => {
+      return {
+        ...state,
+        sourceCode: {
+          ...state.sourceCode,
+          focusLineSpec: focus.sourceLineSpec,
+        },
+      };
+    }
+  ),
+  on(
+    actions.sourceFileRequested,
+    (state: DebuggerState, sourceFileSpec: SourceFileSpec): DebuggerState => {
+      const newState: DebuggerState = {
+        ...state,
+        sourceCode: {
+          ...state.sourceCode,
+          fileContents: state.sourceCode.fileContents.slice(),
         },
       };
+      const fileIndex = findFileIndex(
+        newState.sourceCode.sourceFileList,
+        sourceFileSpec
+      );
+      if (fileIndex >= 0) {
+        newState.sourceCode.fileContents[fileIndex].loadState =
+          DataLoadState.LOADING;
+      } else {
+        throw Error(
```

Comment:
Just checking, is it safe to throw an error in the middle of a Redux/flux cycle?  e.g. if another reducer/effect was supposed to synchronously respond to this action, will it be skipped?  If so, perhaps consider just `console.error`?